### PR TITLE
Add `reverse_graph`

### PR DESCRIPTION
**Pull request by inducer**
cc @matthiasdiener 

- [x] Needs tests

### CODE REVIEW

File: pytools/graph.py [hunk 1]
```diff
@@ -45,13 +46,37 @@
...
+    Reverses a graph.
+
+    :param graph: A :class:`dict` representation of a directed graph, mapping each
+        node to other nodes to which it is connected by edges.
+    :returns: A :class:`dict` representing *graph* with edges reversed.
+    """
+    result: Dict[T, Set[T]] = {}
+
+    for node_key, edges in graph.items():
+        # Make sure every node is in the result even if it has no users
```
Comment (kaushikcfd) [inline pytools/graph.py:69]:
```suggestion
        # Make sure every node is in the result even if it has no successors
```

File: pytools/graph.py [hunk 2]
```diff
@@ -45,13 +46,37 @@
...
 
 
+# {{{ reverse_graph
+
+def reverse_graph(graph: Mapping[T, Collection[T]]) -> Dict[T, FrozenSet[T]]:
+    """
+    Reverses a graph.
+
+    :param graph: A :class:`dict` representation of a directed graph, mapping each
+        node to other nodes to which it is connected by edges.
```
Comment (kaushikcfd) [inline pytools/graph.py:63]: Doesn't explicitly talk about the direction of the edge.

File: pytools/graph.py [hunk 3]
```diff
@@ -45,13 +46,37 @@
...
+    :returns: A :class:`dict` representing *graph* with edges reversed.
+    """
+    result: Dict[T, Set[T]] = {}
+
+    for node_key, edges in graph.items():
```
Comment (kaushikcfd) [inline pytools/graph.py:68]:
```suggestion
    for node_key, successors in graph.items():
```
"edges" makes it sounds like the values are a tuple pair.

**Review by kaushikcfd — APPROVED** [review]
LGTM!
Left some minor readability improvement suggestions.

File: pytools/graph.py (hunk 2 above)
Comment (inducer) [reply pytools/graph.py:63]: I thought about that! I ended up deciding that there was no need to document this, as it's just the reverse of what it was before, no matter which way they were supposed to be pointing initially.

Comment (kaushikcfd) [reply pytools/graph.py:63]: Fair, sg!

**Comment by inducer** [comment]
Thanks all!

**Pull request merged**